Modules/_datetimemodule.c

```diff
@@ -3225,6 +3227,174 @@ date_isoweekday(PyDateTime_Date *self, PyObject *Py_UNUSED(ignored))
+    "datetime.IsoCalendarDate",                         /* tp_name */
+    sizeof(PyDateTime_IsoCalendarDate),                 /* tp_basicsize */
+    0,                                                  /* tp_itemsize */
+    0,                                                  /* tp_dealloc */
+    0,                                                  /* tp_vectorcall_offset */
+    0,                                                  /* tp_getattr */
+    0,                                                  /* tp_setattr */
+    0,                                                  /* tp_as_async */
+    (reprfunc) iso_calendar_date_repr,                  /* tp_repr */
+    0,                                                  /* tp_as_number */
+    0,                                                  /* tp_as_sequence */
+    0,                                                  /* tp_as_mapping */
+    0,                                                  /* tp_hash */
+    0,                                                  /* tp_call */
+    0,                                                  /* tp_str */
+    0,                                                  /* tp_getattro */
+    0,                                                  /* tp_setattro */
+    0,                                                  /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT,                                 /* tp_flags */
+    iso_calendar_date__doc__,                           /* tp_doc */
+    0,                                                  /* tp_traverse */
+    0,                                                  /* tp_clear */
+    0,                                                  /* tp_richcompare */
+    0,                                                  /* tp_weaklistoffset */
+    0,                                                  /* tp_iter */
+    0,                                                  /* tp_iternext */
+    iso_calendar_date_methods,                          /* tp_methods */
+    0,                                                  /* tp_members */
+    iso_calendar_date_getset,                           /* tp_getset */
+    &PyTuple_Type,                                      /* tp_base */
+    0,                                                  /* tp_dict */
+    0,                                                  /* tp_descr_get */
+    0,                                                  /* tp_descr_set */
+    0,                                                  /* tp_dictoffset */
+    0,                                                  /* tp_init */
+    0,                                                  /* tp_alloc */
+    iso_calendar_date_new,                              /* tp_new */
+};
+
+/*[clinic input]
+@classmethod
+datetime.IsoCalendarDate.__new__ as iso_calendar_date_new
+    sequence as seq: object
+[clinic start generated code]*/
+
+static PyObject *
+iso_calendar_date_new_impl(PyTypeObject *type, PyObject *seq)
```

Comment:
If we stick with this version of the implementation, it's probably better to have a constructor that takes 3 C integers rather than a tuple of `PyLong_Long` objects, since the only place this should ever actually be constructed is in the code for `isocalendar()` and in the current form of this implementation, `isocalendar()` is first constructing an unnecessary intermediate `tuple` object, which is then broken apart to create the `IsoCalendarDate` object.

Reply:
https://github.com/python/cpython/blob/0c989db57ec189511dc876973850bbd31c547ec0/Lib/datetime.py#L1216
So it means that we should update the pure python version of `__new__`?